utils/env_preferences/src/apple.rs

```diff
@@ -2,135 +2,203 @@
-            for i in 0..count {
-                // SAFETY: The call to `CFArrayGetValueAtIndex` is safe because we are iterating from 0 to count (`CFArrayGetCount`) which
-                // in itself will always be greater than 0 and less than count ensuring we will not get "out of bounds" error
-                let lang_ptr = unsafe { CFArrayGetValueAtIndex(locale_carr_ref as _, i) };
-
-                if !lang_ptr.is_null() {
-                    // SAFETY: The call to `CFStringGetCStringPtr` because the reference of string we are accessing is not `NULL`
-                    // Returns pointer in O(1) without any memory allocation. This can return NULL so to handle it
-                    let lang_str = unsafe { CFStringGetCStringPtr(lang_ptr as _, 0) };
-
-                    if !lang_str.is_null() {
-                        // SAFETY: A valid `NULL` terminator is present which is a requirement of `from_ptr`
-                        let lang_rust_str = unsafe { CStr::from_ptr(lang_str) }
-                            .to_str()
-                            .unwrap_or("Unknown")
-                            .to_string();
-                        languages.push(lang_rust_str);
-                    }
-
-                    // TODO: In case `lang_str` is `NULL` try retrieving the string using `CFStringGetCString`
-                    // Ref: https://developer.apple.com/documentation/corefoundation/1542721-cfstringgetcstring?language=objc
-                    // Note: Not optimal and may give inconsistent results if buffer is not large enough, must add sanity checks
-                    // whenever implemented
+use core_foundation_sys::{
+    array::{CFArrayGetCount, CFArrayGetValueAtIndex},
+    base::{CFIndex, CFRelease, CFRetain},
+    calendar::{CFCalendarCopyCurrent, CFCalendarCopyLocale, CFCalendarGetIdentifier},
+    locale::{CFLocaleCopyPreferredLanguages, CFLocaleGetIdentifier},
+    string::{
+        kCFStringEncodingUTF8, CFStringGetCString, CFStringGetCStringPtr, CFStringGetLength,
+        CFStringRef,
+    },
+    timezone,
+};
+use libc::c_char;
+use std::ffi::CStr;
+
+use crate::RetrievalError;
+
+// Helper function
+fn get_string(ptr: CFStringRef) -> String {
+    // SAFETY: It returns length of the string, from above conditional statement we ensure
+    // that the `lang_ptr` is not NULL thus making it safe to call
+    let length = unsafe { CFStringGetLength(ptr) as usize };
+
+    let mut c_str_buf = vec![0; length * 4];
+
+    // SAFETY: Safety is ensured by following points
+    // 1. `lang_ptr` is not NULL, checked through conditional statement
+    // 2. `c_str_buf` is large enough and in scope after this call
+    unsafe {
+        CFStringGetCString(
+            ptr,
+            c_str_buf.as_mut_ptr(),
+            c_str_buf.len() as CFIndex,
+            kCFStringEncodingUTF8,
+        );
+    }
+
+    let c_str_buf_u8: Vec<u8> = c_str_buf.iter().map(|&c| c as u8).collect();
```

Comment:
this might or might not optimise out. It's probably better to instantiate `c_str_buf_u8` as `Vec<u8>`, and do a pointer cast to `*mut c_char` when passing it to `CFStringGetCString`

Reply:
Not really sure how to do this tbh. `c_str_buf_u8 ` is already an `Vec<u8>`.

If I am not mistaken are you wishing to have `c_str_buf_u8: Vec<u8>` be the buffer instead of `c_str_buf`?

Reply:
I think this is better no point in having two allocations

Reply:
Sorry I meant instantiate `c_str_buf` as `Vec<u8>`. Currently that gets inferred to `Vec<c_char>`, and this `.collect` reallocates.

Reply:
ah you did this